### PR TITLE
Handle binary AdditionalTexts in the AdditionalTextComparer.

### LINKED ISSUE #59209: Disabled including binaries for generators in latest .net sdk preview
### Describe the bug
When trying to include a binary file for use in a generator, the following error appears:
```
error CS2015: is a binary file instead of a text file
```
The peculiarity is that this appears only at the first attempt to assemble.
There is a bug here either in that including binaries for generators was disabled in this preview version, or in that it only works for the first build attempt.
In the second case, there is a question - why was the use of binary files prohibited? I have a library that generates code to conveniently use EmbeddedResources based on binary AdditionalFiles, I use it in tests - https://github.com/HavenDV/H.Resources.Generator

### To Reproduce
Include any binary file:
```xml
<ItemGroup>
  <AdditionalFiles Include="*.*" />
</ItemGroup>
```

### Further technical details
image
VS 17.1 preview 4.0

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #59209

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs [hunk 1]
```diff
@@ -32,21 +32,42 @@ public bool Equals(AdditionalText? x, AdditionalText? y)
...
+
+                if (xText is null || yText is null || xText.Length != yText.Length)
+                {
+                    return false;
+                }
 
-            if (xText is null || yText is null || xText.Length != yText.Length)
+                return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
+            }
+            catch (InvalidDataException)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs:47]: ❓ Can we remove this exception handler?

File: src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs [hunk 2]
```diff
@@ -32,21 +32,42 @@ public bool Equals(AdditionalText? x, AdditionalText? y)
...
-            var xText = x.GetText();
-            var yText = y.GetText();
+            try
+            {
+                var xText = GetTextOrNullIfBinary(x);
+                var yText = GetTextOrNullIfBinary(y);
+
+                if (xText is null || yText is null || xText.Length != yText.Length)
+                {
+                    return false;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs:42]: ❗ If `xText` and `yText` are both null, then the additional text is _observably_ not changed and can be treated as equal. We should handle this case specifically by returning `true`.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs [hunk 3]
```diff
@@ -32,23 +32,22 @@ public bool Equals(AdditionalText? x, AdditionalText? y)
...
-                return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
+            // If xText and yText are both null, then the additional text is observably not changed
+            // and can be treated as equal.
+            if (xText is null && yText is null)
+            {
+                return true;
             }
-            catch (InvalidDataException)
+
+            if (xText.Length != yText.Length)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs:45]:
```suggestion
            if (xText is null || yText is null || xText.Length != yText.Length)
```

**Review by sharwell — APPROVED** [review]

**Review by jaredpar — APPROVED** [review]

**Comment by jaredpar** [comment]
@chsienki for second approval

**Review by chsienki — APPROVED** [review]
LGTM, thanks!

**Pull request merged**